### DISCUSSION TITLE
Locations within bounds

### DISCUSSION THREAD

**Discussion by RizaHKhan** [clickbar/laravel-magellan#51] [Q&A] [answered]
Hello,

First of all, great library. Really helped me get started on this small map project I'm on.

I'm saving coordinates of an 'Address' like so:

```
      Schema::create('addresses', function (Blueprint $table) {
            $table->id();
            $table->point('geometry');
            $table->text('place_name');
            $table->text('accuracy');
            $table->text('mapbox_id');
            $table->foreignId('user_id')
                ->constrained()
                ->onDelete('cascade');
            $table->timestamps();
        });
```

I use a factory function to see a whole bunch of points:

```
<?php

namespace Database\Factories;

use Clickbar\Magellan\Data\Geometries\Point;
use Illuminate\Database\Eloquent\Factories\Factory;

/**
 * @extends \Illuminate\Database\Eloquent\Factories\Factory<\App\Models\Address>
 */
class AddressFactory extends Factory
{
    /**
     * Define the model's default state.
     *
     * @return array<string, mixed>
     */
    public function definition(): array
    {
        return [
            'geometry' => Point::makeGeodetic(fake()->longitude, fake()->latitude),
            'place_name'  => fake()->streetAddress(),
            'accuracy'    => 'rooftop',
            'mapbox_id'   => fake()->uuid(),
            'user_id'     => 1
        ];
    }
}
```

Great so far, but now I would like to retrieve points within a certain bounds:

```
{
  "_sw": {
    "lng": -160.547111537505,
    "lat": -4.35614275070094
  },
  "_ne": {
    "lng": 74.61284994822947,
    "lat": 75.06938431157567
  }
}
```

Is there a way to do this? Sorry for the really base question...

**Comment by ahawlitschek** [comment]
Hey there,

thanks for the kind words :) 
I will look into your case.

**Comment by ahawlitschek** [comment]
Regarding your question: 
Do you want to restrict the lat/lng while seeding to the bounding box or do you want to query the database after seeding points to only retrieve the points inside your bounding box?

**Comment by studnitz** [comment]
In a project we extended Faker with a Geometry Provider like this:


In `database/factories/FakerProvider/GeometryProvider.php`:

```php
<?php

namespace Database\Factories\FakerProvider;

use Clickbar\Magellan\Data\Geometries\LineString;
use Clickbar\Magellan\Data\Geometries\Point;
use Clickbar\Magellan\Data\Geometries\Polygon;

class GeometryProvider extends \Faker\Provider\Base
{
    private static $frankfurt_latitude_bounds = [
        'min' => 50.079492,
        'max' => 50.141190,
    ];

    private static $frankfurt_longitude_bounds = [
        'min' => 8.634186,
        'max' => 8.730392,
    ];

    public function randomLatitude(): float
    {
        return self::randomFloat(
            min: self::$frankfurt_latitude_bounds['min'],
            max: self::$frankfurt_latitude_bounds['max']
        );
    }

    public function randomLongitude(): float
    {
        return self::randomFloat(
            min: self::$frankfurt_longitude_bounds['min'],
            max: self::$frankfurt_longitude_bounds['max']
        );
    }

    public function randomPoint(int $ssrid = 4326): Point
    {
        return Point::make($this->randomLongitude(), $this->randomLatitude(), null, null, $ssrid);
    }
}
```

Then in `app/Providers/AppServiceProvider.php` `boot()` 

```php
public function boot(): void
{
	// ...

    fake()->addProvider(new \Database\Factories\FakerProvider\GeometryProvider(fake()));

	// ...
}
```

So you could then use in your factory:

```php
return [
    'geometry' => fake()->randomPoint(),
   	 // ...
];
```

This should put you on the right track to make a bouding box specific faker point. :)

**Reply by saibotk** [reply]
We should include such a feature in Magellan :)

**Comment by ahawlitschek** [comment, marked as answer]
We just released a new update which allows using Box2D and Box3D as params for MagellanExpressions. 
So you can now do the following: 

```php
Address::stWhere(ST::contains(Box2D::make(1,1,1,1), 'geometry'))->get();
```
Depending on the SRID set in the Database, this can lead to an Database Error.
e.g.

If your point in the database has SRID=4326:
` Internal error: 7 ERROR:  contains: Operation on mixed SRID geometries (Point, 0) != (Point, 4326)`

This can be avoided by setting an SRID:
```php
Address::stWhere(
   ST::contains(
      ST::setSRID(Box2D::make(1, 1, 1, 1), 4326),
      'geometry'
   )
)->get();
```

>**Note**
> In your case using _contains_ is totally fine, because you just want to check if a point is fully inside the bounding box. If you are using a geometry you might be interested as soon as the geometry intersects with the bounding box, even it is not fully inside it.

**Reply by RizaHKhan** [reply]
This looks like what I'm searching for.
